include/libtorrent/packet_pool.hpp

```diff
@@ -84,26 +86,33 @@ namespace libtorrent
 		// deleter for std::unique_ptr
 		void operator()(packet* p) const
 		{
+			TORRENT_ASSERT(p != nullptr);
 			p->~packet();
 			std::free(p);
 		}
 	};
 
-	inline packet* create_packet(int size)
+	inline packet* create_packet(int const size)
 	{
 		packet* p = static_cast<packet*>(std::malloc(sizeof(packet) + size));
+		if (p == nullptr) aux::throw_ex<std::bad_alloc>();
 		new (p) packet();
+		p->allocated = static_cast<std::uint16_t>(size);
 		return p;
 	}
 
 	using packet_ptr = std::unique_ptr<packet, packet_deleter>;
 
-	template<int ALLOCATE_SIZE>
 	struct TORRENT_EXTRA_EXPORT packet_slab
 	{
-		static const int allocate_size{ ALLOCATE_SIZE };
+		int const allocate_size;
```

Comment:
sure. it's is *possible* for the same optimization to happen now, since this variable is declared const, the compiler can assume it never changes, and constant fold it. whether that happens or not, I don't know.

In either case, all of the other functions in this class do not need to be instantiated per slab size though, that seems to be unnecessary.